Embed Google fonts and local font files in theme

matiasbenedetto opened the pull request:
### What?
This PR adds 2 sections to the plugin that enables the user to: 
- Embed Google Font assets and font face definitions to the active theme.
- Embed local Font assets and font face definitions to the active theme.

### Why?
The objective is to add a simple way to embed font assets in a block theme that after that you can export with the other plugin options.

### How?
These new section work this way:

#### Embed Google fonts:
1. Gets a list of the available Google fonts from Google fonts API. If the API is not working it uses the stored JSON file as a fallback resource.
2. When the user selects a font, all available variants are loaded to generate the preview.
3. After the user selects the desired variants, and clicks the "Add fonts to the theme" button, the font files are written to the theme assets and the font face definitions are added to the theme's `theme.json` file.

#### Embed local font files:
The font file selected by the user is added to the theme assets/fonts folder and the font face definitions are added to the theme's `theme.json` file.

### Screenshots

#### Adding Google Font:
[Screencast from 22-09-22 16:57:49.webm](https://user-images.githubusercontent.com/1310626/191971257-326e3cfc-c4a0-4c1b-b550-14375cd17a23.webm)

#### Adding local font asset:
[Screencast from 29-09-22 15:20:03.webm](https://user-images.githubusercontent.com/1310626/193114801-3d7016c6-bf1c-4c86-aa08-9a8a1e630c8c.webm)

matiasbenedetto committed:
Embed Google fonts in theme

matiasbenedetto committed:
Add Google Fonts: If the new font family already exists in the theme.json font families, add the new font faces to the existing font family

matiasbenedetto committed:
Google fonts: removing not needed dependecies

matiasbenedetto committed:
If the font asset folder can't be written it returns an error

matiasbenedetto committed:
If the API response doesn't have items it throws an error and use the local fonts list

matiasbenedetto committed:
Removing not needed line

matiasbenedetto committed:
Adding the ability to embed local files

matiasbenedetto committed:
font file formar available

matiasbenedetto committed:
adding font file extensions supported hint

paaljoachim commented:
@draganescu @hellofromtonya
I believe you both might want to know about this PR.

pbking commented on admin/js/google-fonts.js:
```diff
@@ -0,0 +1,138 @@
+const API_URL = "https://www.googleapis.com/webfonts/v1/webfonts?key=";
+const API_KEY = "AIzaSyC9-sErFQouy_mc0OqUmkNHEJZiZ-2_3iI"; // The API key is embed here because the API doesn't seem to have limits or costs.
```
Who's API key is this?  

While there might not be limits is this appropriate?  It would be unfortunate to require a user to have to enter this to use the utility but I'm not sure what the rules around distributing the key are.

Can we maybe add a place for a user to paste their own key (and maybe a link/simple instructions to obtain it) instead?  And (if I understand) that key is only needed to get a current list of typefaces... without the key this would still work with the backup JSON right?

matiasbenedetto replied:
The API url was changed to use a proxy to avoid sharing the API key. We are now using this proxy: https://google-fonts-service-proxy.vercel.app/
The old API Key included here was deleted and it is no longer usable.

madhusudhand replied:
I do not think it is best to use vercel as proxy, because of the bandwidth limits that vercel has. It starts failing once the limit reached.

Also Chrome flags it as Phishing site https://google-fonts-service-proxy.vercel.app
<img width="494" alt="image" src="https://user-images.githubusercontent.com/1935113/193239624-0ca061a7-5702-4a2a-acf9-6378362e4c2e.png">

This could be fixed by updating the site, but for the above reason it should be avoided.

madhusudhand replied:
Also, this is adding another layer of dependency on a single user. 
API key of the user + vercel account of the user
For some reason, if vercel decides to suspend user account, this feature fails.

**Update:** I looked into the vercel fair usage policy guidelines. It suggests to not use their service as proxy.
https://vercel.com/docs/concepts/limits/fair-use-policy

matiasbenedetto replied:
It seems unlikely to reach the 100GB traffic quota of the free account. Apart from that, I think this can be considered " Functions that query DBs or APIs" more than a proxy that would be a general purpose proxy. 

> For some reason, if vercel decides to suspend user account, this feature fails.

The feature won't fail because if there is no answer from the API a static fallback JSON included in the plugin is used.

matiasbenedetto replied:
This aspect is not vital for the plugin to work so we can discuss it in another PR. I removed the API fetching code to use just the local JSON file with the Google fonts data until we land a better solution.

pbking commented:
This is (currently) limited to `.ttf` typefaces.  As a future improvement it would be really nice to either default to woff2 (when available) or provide the user with a choice of formats.

pbking committed:
Converted API endpoint from google (requiring an api key) to a proxied API that does not

matiasbenedetto committed:
Revert "Converted API endpoint from google (requiring an api key) to a proxied API that does not"

Because the API response from that URL is not the same as the Google Fonts API and also I does't allow to be called from the browser (CORS).

This reverts commit 8e94c817a274c88c98ed952f48a4f7bdbc6a2b82.

matiasbenedetto committed:
Replacing API url and API key for a proxy to avoid sharing the API key.

matiasbenedetto commented:
> This is (currently) limited to .ttf typefaces. As a future improvement it would be really nice to either default to woff2 (when available) or provide the user with a choice of formats.

The local font asset upload is not limited to .ttf, you can upload .woff and .woff2 also.
The google ones are limited to .ttf because Google fonts just provide the entire font just in .tff. The woff2 versions are provided [only by character subsets](https://fonts.googleapis.com/css?family=Inter) (Latin, Cyrillic, etc.) so it would require adding more files per font or some kind of extra processing.

matiasbenedetto committed:
Adding comment around Google Font API proxy

matiasbenedetto committed:
Embed local font: reading and pre-filling font information when a font file is selected

mikachan commented on admin/class-embed-fonts-in-theme.php:
```diff
@@ -0,0 +1,322 @@
+                            <th scope="row">
+                                <label for="font-file"><?php _e('Font file', 'create-block-theme'); ?></label>
+                                <br>
+                                <small style="font-weight:normal;">.ttf, .woff, .woff2 file extensions supported</small>
```
This should probably be translatable.

mikachan commented on admin/class-embed-fonts-in-theme.php:
```diff
@@ -0,0 +1,322 @@
+				<select name="google-font" id="google-font-id">
+				</select>
+				<br /><br />
+				<p class="hint">Select the font variants you want to include:</p>
```
This should also probably be translatable.

mikachan commented on admin/js/google-fonts.js:
```diff
@@ -0,0 +1,136 @@
+    const selectElement = document.getElementById('google-font-id');
+    const opt = document.createElement("option");
+    opt.value = "";
+    opt.innerHTML = "Select a font ...";
```
This should probably be translatable as well. I assume this would need to be wrapped with the usual translation function in PHP, so perhaps we could add this first value in the select box in `class-embed-fonts-in-theme.php` so we can translate it, and then continue to add the dynamic options here.

mikachan commented on admin/class-embed-fonts-in-theme.php:
```diff
@@ -0,0 +1,322 @@
+                                <label for="font-name"><?php _e('Font name', 'create-block-theme'); ?></label>
+                            </th>
+                            <td>
+                                <input type="text" name="font-name" id="font-name" placeholder="Font Name" required>
```
Placeholder should be translated too.

mikachan commented on admin/class-embed-fonts-in-theme.php:
```diff
@@ -0,0 +1,322 @@
+                                <label for="font-weight"><?php _e('Font weight', 'create-block-theme'); ?></label>
+                            </th>
+                            <td>
+                                <input type="text" name="font-weight" id="font-weight" placeholder="Font Weight" required>
```
Placeholder should be translated too.

mikachan reviewed (commented):
Thanks so much for working on this! I've tested it out and it's working so well.

I think it would be good to update the main readme file with info about this functionality, maybe with its own subheading and section.

This isn't a blocker for this PR and is just my opinion, but I'm not sure about the two links, 'Embed Google font in current Theme' and 'Embed local font in current Theme', in the main left sidebar. I think it would be good to reduce this down to one page so there's only one additional link here ('Embed font in current theme').

matiasbenedetto committed:
Marking a few strings remaining as translatable strings

matiasbenedetto commented:
Thanks @mikachan for the review and for noticing those remaining strings marked as translatable. I addressed your comments in the latest commit.

matiasbenedetto committed:
Removing API fetching for google fonts. Using locally hosted version of the API response

matiasbenedetto committed:
Updating README file

matiasbenedetto commented:
The dynamic API fetching is not vital for the plugin to work so we can discuss it in another PR. I removed the API fetching code to use just the local JSON file with the Google fonts data until we land a better solution. Also I updated the README file to include this new feature.

mikachan reviewed (approved):
> The dynamic API fetching is not vital for the plugin to work so we can discuss it in another PR.

Good idea 👍 I've tested this PR again with your latest changes and everything is still working well for me.

> Also I updated the README file to include this new feature.

Ace, looks great!

matiasbenedetto commented:
Thanks everyone for the feedback! I'm merging this
